controller/src/main/java/io/pravega/controller/server/SegmentHelper.java

```diff
@@ -88,64 +103,77 @@ public SegmentHelper(final ConnectionFactory clientCF, HostControllerStore hostS
-            }
+    @SneakyThrows(ConnectionFailedException.class)
+    private Boolean transformReplyToBoolean(Reply reply, Class[] classArray) {
+        if (isReplyWanted(reply, classArray)) {
+            return true;
+        }
 
-            @Override
-            public void processingFailure(Exception error) {
-                log.error(requestId, "CreateSegment {} threw exception", qualifiedStreamSegmentName, error);
-                handleError(error, result, type);
-            }
+        closeConnection(reply);
+        if (reply instanceof WireCommands.NoSuchSegment) {
+            throw new NoSuchSegmentException(reply.toString());
+        } else if (reply instanceof WireCommands.WrongHost) {
+            throw new ConnectionFailedException(reply.toString());
+        } else {
+            throw new ConnectionFailedException("Unexpected reply of " + reply + " when expecting a "
+                    + classArray[0].getName());
+        }
+    }
 
-            @Override
-            public void authTokenCheckFailed(WireCommands.AuthTokenCheckFailed authTokenCheckFailed) {
-                result.completeExceptionally(
-                        new WireCommandFailedException(new AuthenticationException(authTokenCheckFailed.toString()),
-                                type, WireCommandFailedException.Reason.AuthFailed));
-            }
-        };
+    public CompletableFuture<Boolean> createSegment(final String scope,
+                                                    final String stream,
+                                                    final long segmentId,
+                                                    final ScalingPolicy policy,
+                                                    String controllerToken,
+                                                    final long clientRequestId) {
+        final String qualifiedStreamSegmentName = getQualifiedStreamSegmentName(scope, stream, segmentId);
+        final Controller.NodeUri uri = getSegmentUri(scope, stream, segmentId);
+        //final long requestId = (clientRequestId == RequestTag.NON_EXISTENT_ID) ? idGenerator.get() : clientRequestId;
+        final long requestId = (clientRequestId == RequestTag.NON_EXISTENT_ID) ? idGenerator.get() : idGenerator.get() + clientRequestId;
 
+        RawClient connection = getConnection(ModelHelper.encode(uri));
         Pair<Byte, Integer> extracted = extractFromPolicy(policy);
 
-        WireCommands.CreateSegment request = new WireCommands.CreateSegment(requestId, qualifiedStreamSegmentName,
-                extracted.getLeft(), extracted.getRight(), controllerToken);
-        sendRequestAsync(request, replyProcessor, result, ModelHelper.encode(uri));
-        return result;
+        Class[] expectedArray = {WireCommands.SegmentCreated.class,
+                WireCommands.SegmentAlreadyExists.class};
+
+        return connection.sendRequest(requestId, new WireCommands.CreateSegment(requestId, qualifiedStreamSegmentName,
```

Comment:
we are losing all the logging specific to the response received. 
We should find a way to add appropriate logs.